### ISSUE TITLE
charging not stopped by ev_smart_charging integration

### ISSUE THREAD

**Issue by ronajon** [jonasbkarlsson/ev_smart_charging#161]
## Version of the custom_component

Version 1.7.0

## Configuration
Home Assistant Core
Tesla custom integration
electricity prices come from the Noordpool integration
EV SOC entity : sensor.mario_kart_battery
EV Target SOC entity: number.mario_kart_charge_limit (also tried with blank value)
Charger control switch: switch.mario_kart_charger

## Describe the bug
when charging starts, the ev_smart_charging integration does not stop the charging and wait until the lowest prices.
when the charger cable is plugged in the car, it starts charging directly and only stops when the batteries are full. 
Charging can be interrupted by clicking the button "manually stop charging", then the ev_smart_charging integration waits until the lowest price and starts charging. 

## Expected behaviour
when the cable is plugged in, the charging starts. 
the ev_smart_charging integratoin notices that the current time does not have the lowest price and stops the charging. 
charging is resumed after the lowest price is met. 

## Debug log
debug log can be found here [https://controlc.com/7cc9dc9f](url)
the cable was connected to the car at 2023-03-14 18:36:00 (line 2482 of the log file )

```text

Add your logs here.

```

**Comment by jonasbkarlsson** [comment]
Thanks for providing the debug log. Very helpful!

At 2023-03-14 18:36, the state of this integration was "off"
`2023-03-14 18:36:10.898 DEBUG (MainThread) [custom_components.ev_smart_charging.coordinator] current_value = False`

So if charging started when you connected the car, there was something else that was wrong. I noticed that the Tesla integration failed earlier when the charging was supposed to end when the car got fully charged at 04:55.

```
2023-03-14 04:54:51.409 DEBUG (MainThread) [custom_components.ev_smart_charging.coordinator] Before service call switch.turn_off: switch.mario_kart_charger
2023-03-14 04:55:11.150 ERROR (MainThread) [homeassistant.core] Error executing service: <ServiceCall switch.turn_off (c:01GVF43B4HCCZGRZ4XEQ2KVT75): entity_id=['switch.mario_kart_charger']>
  File "/srv/homeassistant/lib/python3.10/site-packages/teslajsonpy/connection.py", line 209, in __open
    raise TeslaException(resp.status_code, "vehicle_unavailable")
teslajsonpy.exceptions.TeslaException: vehicle_unavailable 
```

So my guess would be that your problem was due to the Tesla integration. Maybe you need to reload it, or restart Home Assistant to make it work again?

**Comment by jonasbkarlsson** [comment]
Ah, now I realized that your Tesla integration controls the car, and not a charger. For this to work I think you need to make a few automations.

- One automation that sets `switch.ev_smart_charging_ev_connected` when connected to your charger would probably be good to have.
- And then you need an automation that stops the charging, if charging has started and `sensor.ev_smart_charging_charging` is off.

And these two automations needs be able to distinguish between charging at home and charging at other places.

**Comment by jonasbkarlsson** [comment]
Hmm. I was thinking a little more about this, and I think I can modify the integration to do the second task, to turn off charging directly after the car has been connected to the charger. 

However, the first one, to add an automation that sets `switch.ev_smart_charging_ev_connected` will be needed.

**Comment by WhippingBoy01** [comment]
I'm facing the same with my Tesla, it has a full API integrated into home assistant. so we can provide 

binary_sensor.Tesla_charger PLUGGED IN| UNPLUGGED 
binary_sensor.herbie_charging CHARGING|NOT CHARGING

https://github.com/alandtse/tesla

I also have a Tesla wall connector integrated to HA with 
binary_sensor.tesla_wall_connector_vehicle_connected ON|OFF


I would use the above, to avoid the HA blocking the charging in the wild at a super charinger ;-) 

Maybe it could make sense to add an optional field in the configuration window:
EV connected:
binary_sensor.tesla_wall_connector_vehicle_connected

Coffee your way ;-)

**Comment by WhippingBoy01** [comment]
I use this automation as a temporary fix

alias: Tesla stop charging 
description: ""
trigger:
  - platform: state
    entity_id:
      - binary_sensor.tesla_wall_connector_vehicle_connected
    from: "off"
    to: "on"
condition:
  - condition: state
    entity_id: sensor.ev_smart_charging_charging
    state: "off"
action:
  - service: switch.turn_off
    data: {}
    target:
      entity_id: switch.Tesla_charger
mode: single

**Comment by jonasbkarlsson** [comment]
> I also have a Tesla wall connector integrated to HA with binary_sensor.tesla_wall_connector_vehicle_connected ON|OFF
> 
> I would use the above, to avoid the HA blocking the charging in the wild at a super charinger ;-)
> 
> Maybe it could make sense to add an optional field in the configuration window: EV connected: binary_sensor.tesla_wall_connector_vehicle_connected

My current thinking is to add a configuration option for the user to be able to state that control of charging is done via an EV integration. And only if this option is selected, the integration will turn off chariging when the EV is connected to the charger.   

The reason for this is to make sure that there are no users that today have fully functional configuration will experience "the HA blocking the charging in the wild at a super charinger" due to poorly configured automation for the `switch.ev_smart_charging_ev_connected `.

As for making the integration using `binary_sensor.tesla_wall_connector_vehicle_connected` as input to "EV connected", in general I have so far avoided making the integration too specific. There are many EVs and many chargers that might make it unreasonable to handle each of them. However, I do have some ideas about making the integration directly interact with more chargers, and maybe similar ideas could be used to also handle EV Connected.

**Comment by WhippingBoy01** [comment]
above should only have been mentioned as an example "Maybe it could make sense to add an optional field in the configuration window: EV connected: binary_sensor.tesla_wall_connector_vehicle_connected". 

Tesla's have a location tracker as well.
device_tracker.Tesla_location_tracker

It could be used as a workaround, I did update the script above with it.

keeping the integration as vanilla/generic sounds like a very good idea ;-)

**Comment by ronajon** [comment]
> Ah, now I realized that your Tesla integration controls the car, and not a charger. For this to work I think you need to make a few automations.
> 
>     * One automation that sets `switch.ev_smart_charging_ev_connected` when connected to your charger would probably be good to have.
> 
>     * And then you need an automation that stops the charging, if charging has started and `sensor.ev_smart_charging_charging` is off.
> 
> 
> And these two automations needs be able to distinguish between charging at home and charging at other places.
> 

the switch.ev_smart_charging_ev_connected is always on. i have not yet found how or why it would tutn off (except the one time i manually tuned it off)

creating the automations myself is not an issue. is there info available about the flow of the integration? meaning how the integration determines when to start charging or not ?

ps. i don't have a Tesla wall connector, and agree that keeping the integration should be kept as generic as possible

**Comment by pbsiim** [comment]
I have an VW ID.3. The API seems a little byggy/skow. The manuel press Start/stop does not always work. The same goes for the planned charging start/stop.
When using a call service for stop/start buttons it works like a charm every time. Im not sure why?

```
action: call-service
      service: volkswagen_we_connect_id.volkswagen_id_start_stop_charging
      data:
        vin: WVWZZZxxxxxxx
        start_stop: start
```
My question is if there are any good way to use the same method for the planned start/stop?

**Comment by jonasbkarlsson** [comment]
You can use a service in this way for starting a planned charging:
```
alias: EV Smart Charging - Start
description: ""
mode: single
trigger:
  - platform: state
    entity_id:
      - sensor.ev_smart_charging_charging
    from: "off"
    to: "on"
condition: []
action:
  - service: volkswagen_we_connect_id.volkswagen_id_start_stop_charging
    data:
      vin: WVWZZZxxxxxxx
      start_stop: start
```

But what is it that doesn't work with your current confguration? How have you configured the integration to start/stop charging?

**Comment by pbsiim** [comment]
> You can use a service in this way for starting a planned charging:
> 
> ```
> alias: EV Smart Charging - Start
> description: ""
> mode: single
> trigger:
>   - platform: state
>     entity_id:
>       - sensor.ev_smart_charging_charging
>     from: "off"
>     to: "on"
> condition: []
> action:
>   - service: volkswagen_we_connect_id.volkswagen_id_start_stop_charging
>     data:
>       vin: WVWZZZxxxxxxx
>       start_stop: start
> ```
> 
> But what is it that doesn't work with your current confguration? How have you configured the integration to start/stop charging?



Thanks for the fast response. I will not go into detail in how i did the automation, but will follow your example as it's much more clean and simple. I was not aware of the "correct" method.
I'm pretty sure I’m all good now, the only “issue” I have is the button from the example does not always work.

`entity: button.ev_smart_charging_manually_stop_charging`

However im using mushroom cards and it works just fine like this.


```
tap_action:
      action: call-service
      service: volkswagen_we_connect_id.volkswagen_id_start_stop_charging
      target: {}
      data:
        vin: WVWZZZxxxx
        start_stop: start
```

**Comment by jonasbkarlsson** [comment]
At least partly fixed in release [v1.9.0](https://github.com/jonasbkarlsson/ev_smart_charging/releases/tag/v1.9.0). If there still are issues, please create a new issue.

Note that there now is a configuration item that will tell the integratation that it is an EV integration as the Tesla custom integration (and not a charger integration) that controls charging. Please update this in your configuration.
